lib/charms/tls_certificates_interface/v3/tls_certificates.py

```diff
@@ -450,6 +450,10 @@ class ProviderCertificate:
     chain: List[str]
     revoked: bool
 
+    def chain_as_pem(self) -> str:
```

Comment:
To my understanding,
- `chain_as_pem` should be a private utility method elsewhere; and
- `.certificate` should be populated with the output of `chain_as_pem` to begin with. Or at the very least with the CA's own signature.

I'm still confused though, because [the following curl](https://github.com/canonical/traefik-k8s-operator/blob/a71129a0fa8e26fda56b96f94a805364155d37b4/tests/integration/test_tls.py#L99) already works as I expect:
```python
        cmd = [
            "curl",
            "-s",
            "--fail-with-body",
            "--resolve",
            f"{mock_hostname}:443:{traefik_app_ip}",
            "--capath",
            certs_dir,
            "--cacert",
            cert_path,
            endpoint,
        ]
```

Reply:
The reason your test works is that you are using `self-signed-certificates`, and in that case the whole chain consists of `Certificate -> CA`. When using Let's Encrypt, the certificate is signed by an intermediate CA, and the whole chain looks like: `Certificate -> Intermediate CA -> CA`.

The thing I am not sure with changing what `.certificate` returns is that it might break some existing applications. Depending on the workload, they might need the certificates in different format. I think the most frequent one is having the certificate and intermediates in one file and the CA in another, but I don't want to assume.

That is the reason why I was leaning towards public methods that make it easy for charm developers to get the format they need from what we have in the library.

Reply:
We discussed this internally and we will go with this approach for now, and we might reconsider when we have enough breaking changes to justify a new major version.